packages/apollo-gateway/src/index.ts

```diff
@@ -78,32 +98,33 @@ export class ApolloGateway implements GraphQLService {
     loglevelDebug(this.logger);
 
     // And also support the `debug` option, if it's truthy.
-    if (config.debug === true) {
+    if (this.config.debug === true) {
       this.logger.enableAll();
     }
 
-    if (isLocalConfig(config)) {
-      this.createSchema(config.localServiceList);
+    if (isLocalConfig(this.config)) {
+      this.createSchema(this.config.localServiceList);
     }
 
     this.initializeQueryPlanStore();
   }
 
-  public async load() {
+  public async load(engineConfig?: EngineConfig) {
+    this.engineConfig = engineConfig;
     if (!this.isReady) {
       this.logger.debug('Loading configuration for Gateway');
       const [services] = await this.loadServiceDefinitions(this.config);
       this.logger.debug('Configuration loaded for Gateway');
       this.createSchema(services);
     }
 
-    return { schema: this.schema, executor: this.executor };
+    return { schema: this.schema!, executor: this.executor };
```

Comment:
I wonder if there is something that we can restructure in the flow that would allow this to get narrowed on its own?  Maybe using a `private getSchema` method could help?  Trying to avoid those non-null assertions and would love to explore options!

Reply:
The invariants are: “if `this.isReady` is true, then `this.schema` is defined”, and “after calling `this.createSchema`, `this.schema` is defined.”

As for encoding that in TS... the top level check could be on schema instead of isReady, and createSchema could return the created schema, which we then return?